build_runner/lib/src/generate/performance_tracker.dart

```diff
@@ -143,19 +218,60 @@ class BuilderActionTracker extends TimeTracker
   }
 }
 
+/// No-op instance of [BuilderActionTracker] which does nothing and throws an
+/// unimplemented error for everything but [track], which delegates directly to
+/// the wrapped function.
+///
+/// Use the [BuilderActionTracker.noOp] factory to get an instance.
+class _NoOpBuilderActionTracker extends Object
+    with _NoOpTimeTracker
+    implements BuilderActionTracker {
+  static final _NoOpBuilderActionTracker _sharedInstance =
+      new _NoOpBuilderActionTracker();
+
+  @override
+  Builder get builder => throw new UnimplementedError();
+
+  @override
+  Duration get duration => throw new UnimplementedError();
+
+  @override
+  Iterable<BuilderActionPhasePerformance> get phases =>
+      throw new UnimplementedError();
+
+  @override
+  AssetId get primaryInput => throw new UnimplementedError();
+
+  @override
+  FutureOr<T> track<T>(FutureOr<T> runPhase(), String label) => runPhase();
+}
+
 /// Tracks the [Timings] of an indivual task.
 ///
 /// These represent a slice of the [BuilderActionPerformance].
-class BuilderActionPhaseTracker extends TimeTracker
+class BuilderActionPhaseTracker extends Object
+    with _TimeTrackerImpl
     implements BuilderActionPhasePerformance {
   @override
   final String label;
 
   BuilderActionPhaseTracker(this.label);
 }
 
-/// Internal base class for tracking the [Timings] of an arbitrary operation.
-class TimeTracker implements Timings {
+/// Interface for tracking the [Timings] of an operation using the [start] and
+/// [stop] methods.
+abstract class TimeTracker implements Timings {
+  factory TimeTracker() => new _TimeTrackerImpl();
+  factory TimeTracker.noOp() => _NoOpTimeTracker.sharedInstance;
+
+  void start();
+  void stop();
+}
+
+/// Implementation of a real [TimeTracker].
+///
+/// Use [TimeTracker] factory to get an instance.
+class _TimeTrackerImpl implements TimeTracker {
```

Comment:
IMO the default TimeTracker class should be this and the NoOp version should extend it.

Reply:
My only problem with that is if we add new methods/fields, the NoOp version would end up inheriting the real implementation. With it set up this way we explicitly have to implement anything new with either a no-op or throw which is a bit safer long term.